### DISCUSSION TITLE
How do you know if a message has been succesfully sent?

### DISCUSSION THREAD

**Discussion by fercom** [gabrieldwight/Whatsapp-Business-Cloud-Api-Net#39] [Q&A] [answered]
When I call:
`var results = await _whatsAppBusinessClient.SendImageAttachmentTemplateMessageAsync(imageTemplateMessage);`

I am receiving something like this in the result variable:
`{"messaging_product":"whatsapp","contacts":[{"input":"XXXXXXXXXX","wa_id":"XXXXXXXXXX"}],"messages":[{"id":"wamid.HBgNNTIxNzcxMTM2NzU4MRUCABEYEjg4Q0YzNDMyMzZCMkExM0Q4RAA="}]}`

does the "wamid" means that the message has been sent?

I am asking because I just sent a large volume of messages and saved the responses to my database, but I want to know how many were succesfully sent

**Comment by gabrieldwight** [comment, marked as answer]
The "wamid" is just the message identifier for the sent message. If you want to know if the message is sent, the message status is sent through the message webhook since it will be triggered every time the user views it. The message statuses also include failed or deleted. Below is the sample message status json sent to the webhook since the response will be dynamic changing to different objects. You can compare the "wamid" with the id field in the statuses object and then do a database update.

Message status json
```JSON
{
    "object": "whatsapp_business_account",
    "entry": [
        {
            "id": "8856996819413533",
            "changes": [
                {
                    "value": {
                        "messaging_product": "whatsapp",
                        "metadata": {
                            "display_phone_number": "",
                            "phone_number_id": ""
                        },
                        "statuses": [
                            {
                                "id": "wamid.gBGGFlCGg0cvAglAxydbAoy-gwNo",
                                "status": "sent", // delivered, failed, deleted or read
                                "timestamp": "",
                                "recipient_id": ""
                            }
                        ]
                    },
                    "field": "messages"
                }
            ]
        }
    ]
}
```

```C#
var changesResult = messageReceived["entry"][0]["changes"][0]["value"];

                if (changesResult["statuses"] != null)
                {
                    var messageStatus = Convert.ToString(messageReceived["entry"][0]["changes"][0]["value"]["statuses"][0]["status"]);

                    if (messageStatus.Equals("sent")) // one tick
                    {
                        var messageStatusReceived = JsonConvert.DeserializeObject<UserInitiatedMessageSentStatus>(Convert.ToString(messageReceived)) as UserInitiatedMessageSentStatus;
                        var messageStatusResults = new List<UserInitiatedStatus>(messageStatusReceived.Entry.SelectMany(x => x.Changes).SelectMany(x => x.Value.Statuses));
                        _logger.LogInformation(JsonConvert.SerializeObject(messageStatusResults, Formatting.Indented));

                        return Ok(new
                        {
                            Message = $"Message Status Received: {messageStatus}"
                        });
                    }

                    if (messageStatus.Equals("delivered")) // double tick
                    {
                        var messageStatusReceived = JsonConvert.DeserializeObject<UserInitiatedMessageDeliveredStatus>(Convert.ToString(messageReceived)) as UserInitiatedMessageDeliveredStatus;
                        var messageStatusResults = new List<UserInitiatedMessageDeliveryStatus>(messageStatusReceived.Entry.SelectMany(x => x.Changes).SelectMany(x => x.Value.Statuses));
                        _logger.LogInformation(JsonConvert.SerializeObject(messageStatusResults, Formatting.Indented));

                        return Ok(new
                        {
                            Message = $"Message Status Received: {messageStatus}"
                        });
                    }

                    if (messageStatus.Equals("read")) // blue tick
                    {
                        return Ok(new
                        {
                            Message = $"Message Status Received: {messageStatus}"
                        });
                    }
                }
```

**Reply by kasunf** [reply]
If you are using temporary token,

Try following.

Send Test Message using https://developers.facebook.com/
Token will change every 23 hours , however if you refresh the developer token page its changing too.
Reply to test msg.
Then try sending msg programmatically.
Phone nos without + worked for me.

**Comment by fercom** [comment]
Thanks @gabrieldwight  that is what I was looking for

**Comment by theCJMan** [comment]
> When I call: `var results = await _whatsAppBusinessClient.SendImageAttachmentTemplateMessageAsync(imageTemplateMessage);`
> 
> I am receiving something like this in the result variable: `{"messaging_product":"whatsapp","contacts":[{"input":"XXXXXXXXXX","wa_id":"XXXXXXXXXX"}],"messages":[{"id":"wamid.HBgNNTIxNzcxMTM2NzU4MRUCABEYEjg4Q0YzNDMyMzZCMkExM0Q4RAA="}]}`
> 
> does the "wamid" means that the message has been sent?
> 
> I am asking because I just sent a large volume of messages and saved the responses to my database, but I want to know how many were succesfully sent

Hey guys, 
My question relates to this topic. I am using this: 
var result = await sendMessageController.SendTemplate_video_ParameterAsync(sendWhatsAppPayload);

and the return or result is of type Task<ActionResult<WhatsAppResponse>>

How do I get the wamid from result?

thanx

**Reply by gabrieldwight** [reply]
```C#
var messages = result.Messages;
string messageId = results.Messages[0].Id; \\ for waim.XXXX depends how many id are existing in the array

return Ok(messageId) or return Ok(result.Messages)
```

**Reply by theCJMan** [reply]
When or Why might there be more than 1 message or IDs

**Reply by gabrieldwight** [reply]
If the message is conversation-based it mostly happens in the webhook if it is either initiated by the business or a user. It's done on Whatsapp end.

**Reply by theCJMan** [reply]
ok, so from the Send (us using APIs) side not the Receive side (WebHook) there should not be more than 1 Message?
you see, I am also writing code to do bulk sends:
1. Reading a CSV
2. Send 1 WhatsApp per line
3. If the message went out, i need to capture this WAMId to keep track in the Webhook if a client is responding to my message sent..

**Reply by theCJMan** [reply]
You see I am thinking of having this method added, but not sure if it will be beneficial 

 public string GetWAMId(WhatsAppResponse whatsAppResponse)
 {
		// This is to try and get the WAMID

		StringBuilder msgIDsBuilder = new StringBuilder();

		foreach (Message msg in whatsAppResponse.Messages)
		{
			msgIDsBuilder.Append(msg.Id);
			msgIDsBuilder.Append(", ");
		}

		return msgIDsBuilder.ToString();
	}

**Comment by gabrieldwight** [comment]
> ok, so from the Send (us using APIs) side not the Receive side (WebHook) there should not be more than 1 Message? you see, I am also writing code to do bulk sends:
> 
> 1. Reading a CSV
> 2. Send 1 WhatsApp per line
> 3. If the message went out, i need to capture this WAMId to keep track in the Webhook if a client is responding to my message sent..


By default, it is 1 message from the cloud API side. If you are sending messages in bulk you must create your implementation to handle batch processing and tracking their respective WAMId.

**Comment by theCJMan** [comment]
Yes, so after every send i grab the WAMId and save it to that record.

I will then, when my app is live, focus on the web hooks to check if the message was delivered, red, not a WhatsApp number, etc..



⁣Get BlueMail for Android ​
